File: java/libgsc/src/actor/Gusr.java
package actor;

import misc.Log;
import misc.Misc;
import misc.Net;
import net.Pro;
import stmp.Stmp;
import stmp.StmpDec;
import stmp.StmpEnc;
import stmp.StmpNode;
import stmp.StmpPdu;

import com.google.protobuf.GeneratedMessage.Builder;
import com.google.protobuf.Message;
import com.google.protobuf.ProtocolMessageEnum;

import core.Cb;
import core.Cfg;
import core.Gn2htrans;
import core.Gsc;

/**
 *
 * libgsc N2H连接上的用户数据抽象.
 * 
 * @Created on: 2015年1月10日 下午12:27:18
 * @Author: xuzewen
 * 
 */
public abstract class Gusr extends Actor
{
	/** 关联的N2H网络连接. */
	public N2H n2h = null;
	/** Gusr唯一标识. */
	public long uid;

	public Gusr(long uid)
	{
		super(ActorType.ITC);
		this.wk = Gsc.hashWk(uid);
		this.uid = uid;
	}

	/** ---------------------------------------------------------------- */
	/**                                                                  */
	/**  */
	/**                                                                  */
	/** ---------------------------------------------------------------- */
	/** 连接断开事件. */
	public abstract void disc();

	/** 消息出栈时的加密函数. */
	public abstract byte[] encryp(byte[] by);

	/** 消息入栈时的解密函数. */
	public abstract byte[] decryp(byte[] by);

	/** ---------------------------------------------------------------- */
	/**                                                                  */
	/**  */
	/**                                                                  */
	/** ---------------------------------------------------------------- */

	/** 处理长连接上的STMP-BEGIN. */
	public final void procBegin0(byte[] sec, Gn2htrans gt)
	{
		StmpNode root = StmpDec.unpack(this.decryp(sec));
		if (root == null)
		{
			this.close();
			return;
		}
		Short cmd = StmpDec.getShort(root, Stmp.STMP_TAG_CMD);
		if (cmd == null)
		{
			if (Log.isDebug())
				Log.debug("missing required field: STMP_TAG_CMD");
			this.close();
			return;
		}
		Cb cb = Gsc.evns.get(cmd.intValue());
		if (cb == null)
		{
			if (Log.isDebug())
				Log.debug("can not found call back for this cmd: %04X", cmd);
			this.close();
			return;
		}
		if (cb.gusr != true || cb.beginName == null) /* 不应该在这里收到. */
		{
			if (Log.isDebug())
				Log.debug("it`s an unexpected cmd: %s", cb.cmd);
			this.close();
			return;
		}
		byte[] pb = StmpDec.getBin(root, Stmp.STMP_TAG_DAT);
		if (pb == null) /* 为简单起见(避免在业务处理中频繁判断BEGIN是否为空), BEGIN中的pb对象为必选. */
		{
			if (Log.isDebug())
				Log.debug("missing required field: STMP_TAG_DAT");
			this.close();
			return;
		}
		try
		{
			gt.begin = cb.newBeginMsg(pb);
			gt.extbegin = StmpDec.getBin(root, Stmp.STMP_TAG_ATT);
			gt.cmd = cb.cmd;
			if (Log.isTrace())
			{
				Log.trace("\n  <-- BEGIN(%s): {%s}, END(%s), UID: %d, CMD: %s, TID: %08X, PEER: %s\n", //
						cb.beginName, Misc.pb2str(gt.begin), cb.endName, this.uid, gt.cmd, gt.tid, Net.getRemoteAddr(this.n2h.sc));
			}
			cb.cb.invoke(null, this, gt, gt.begin);
		} catch (Exception e)
		{
			if (Log.isDebug())
				Log.debug(Log.trace(e));
			this.close();
			return;
		}
	}

	/** 处理短连接上的STMP-BEGIN. */
	public final void procBegin1()
	{
		Log.warn("unsupported.");
	}

	/** END出栈. */
	public final void sendEnd0(Gn2htrans gt)
	{
		if (Log.isTrace())
		{
			Cb cb = Gsc.evns.get(gt.cmd.getNumber());
			if (cb != null)
			{
				Log.trace("\n  --> END(%s): {%s}, BEGIN(%s): {%s}, UID: %d, RET: %04X, CMD: %s, TID: %08X, PEER: %s\n", //
						cb.endName, gt.end.end == null ? "NULL" : Misc.pb2str(gt.end.end), //
						cb.beginName, gt.begin == null ? "NULL" : Misc.pb2str(gt.begin),//
						this.uid, gt.end.ret, gt.cmd, gt.tid, this.n2h == null ? "NULL" : Net.getRemoteAddr(this.n2h.sc));
			} else
				Log.fault("it`s a bug, cmd: %s", gt.cmd);
		}
		byte sec[] = this.encryp(Pro.pkgEndSecDat(gt.end.ret, gt.end.end, gt.end.extend)); /* 打包加密块. */
		StmpPdu end = new StmpPdu(sec.length);
		StmpEnc.addBin(end, Stmp.STMP_TAG_DAT, sec);
		StmpEnc.addInt(end, Stmp.STMP_TAG_DTID, gt.tid);
		StmpEnc.addTag(end, Stmp.STMP_TAG_TRANS_END);
		this.n2h.future(v -> {
			this.n2h.send(end.bytes()); /* 前转到N2H的线程出栈. */
			gt.finish();
		});
	}

	/** END出栈(短连接). */
	public final void sendEnd1()
	{
		Log.warn("unsupported.");
		return;
	}

	/** UNI出栈. */
	public final void senUni(ProtocolMessageEnum cmd, Message uni, byte[] ext)
	{
		if (Log.isTrace())
		{
			Cb cb = Gsc.evns.get(cmd.getNumber());
			if (cb != null)
			{
				Log.trace("\n  --> UNI(%s): {%s}, UID: %d, CMD: %s, PEER: %s\n", //
						cb.uniName, uni == null ? "NULL" : Misc.pb2str(uni),//
						this.uid, cmd, this.n2h == null ? "NULL" : Net.getRemoteAddr(this.n2h.sc));
			} else
				Log.fault("it`s a bug, cmd: %s", cmd);
		}
		byte[] by = this.encryp(Pro.pkgUniSecDat(cmd, uni, ext)); /* 加密. */
		StmpPdu pdu = new StmpPdu(by.length);
		StmpEnc.addBin(pdu, Stmp.STMP_TAG_DAT, by);
		StmpEnc.addTag(pdu, Stmp.STMP_TAG_TRANS_UNI);
		this.n2h.future(v -> this.n2h.send(pdu.bytes())); /* 前转到N2H的线程出栈. */
		if (Gsc.ec != null)
			Gsc.ec.gn2hSendUniOnGusr(this, cmd, uni, ext);
	}

	/** 是否在线. */
	public final boolean isOnline()
	{
		return this.n2h != null;
	}

	/** 强制断开连接(延迟操作). */
	public final void lazyKick()
	{
		N2H n2h = this.n2h;
		n2h.future(v -> n2h.lazyClose());
		this.n2h = null;
	}

	/** 踢下线(将导致连接被延迟关闭). */
	public final void kick(ProtocolMessageEnum cmd, Builder<?> b)
	{
		Message msg = b == null ? null : b.build();
		byte[] by = this.encryp(Pro.pkgUniSecDat(cmd, msg, null)); /* 加密. */
		StmpPdu pdu = new StmpPdu(by.length);
		StmpEnc.addBin(pdu, Stmp.STMP_TAG_DAT, by);
		StmpEnc.addTag(pdu, Stmp.STMP_TAG_TRANS_UNI);
		N2H n2h = this.n2h;
		n2h.future(v -> {
			n2h.send(pdu.bytes()); /* 消息出栈. */
			n2h.lts = System.currentTimeMillis() + Cfg.libgsc_n2h_lazy_close;
			n2h.lazyClose(); /* 延迟关闭. */
			n2h.gusr = null;
		});
		this.n2h = null;
		if (Gsc.ec != null)
			Gsc.ec.gn2hSendUniOnGusr(this, cmd, msg, null);
	}

	/** 踢下线, 无消息提示(连接被立即关闭). */
	public final void kickNoMsg()
	{
		N2H n2h = this.n2h;
		n2h.future(v -> {
			n2h.getGworker().removeActorNet(n2h); /* 关闭连接. */
			/**
			 *
			 * 这里为了达到静默关闭目的, 没调用n2h->close();
			 *
			 * */
		});
		this.n2h = null;
	}

	/** 关闭连接, 断开与N2H之间的关联. */
	private final void close()
	{
		N2H n2h = this.n2h;
		n2h.future(v -> {
			n2h.getGworker().removeActorNet(n2h);
			n2h.evnDis();
		});
	}
}

File: c/libgsc/actor/Gusr.cpp
/*
 * Gusr.cpp
 *
 *  Created on: Jan 31, 2015
 *      Author: root
 */

#include "Gusr.h"
#include "../core/Cfg.h"
#include "../core/Gsc.h"
#include "../net/Lx.h"

Gusr::Gusr(ullong uid, const char* name) :
		Actor(ACTOR_ITC, name, Gsc::hashWk(uid))
{
	this->uid = uid;
	this->n2h = NULL;
}

/** 处理长连接上的STMP-BEGIN. */
void Gusr::procBegin0(uchar* dat, uint len, Gn2htrans* gt)
{
	uint out;
	uchar* msg = this->decryp(dat, len, &out); /** 将length身后的字段解密. */
	stmp_node root;
	if (stmpdec_unpack(msg, out, &root) != 0)
	{
		loop: //
		N2H* n2h = this->n2h;
		n2h->future([n2h, gt] /** 前往N2H关闭连接. */
		{
			n2h->close();
			delete gt; /** 删除事务. */
		});
		free(msg);
		return;
	}
	if (stmpdec_get_short(&root, STMP_TAG_CMD, &gt->cmd) != 0)
	{
		LOG_DEBUG("missing required field: STMP_TAG_CMD")
		stmpdec_free(&root);
		goto loop;
	}
	auto it = Gsc::evns.find(gt->cmd);
	if (it == Gsc::evns.end()) /* 不支持的命令字 . */
	{
		LOG_DEBUG("can not found call back for this cmd: %04X", gt->cmd)
		stmpdec_free(&root);
		goto loop;
	}
	Cb* cb = it->second;
	if (cb->gusr != true || cb->begin == NULL) /** 不应该在这里收到. */
	{
		LOG_DEBUG("it`s an unexpected cmd: %04X", it->first)
		stmpdec_free(&root);
		goto loop;
	}
	uchar* pb = stmpdec_peek_bin(&root, STMP_TAG_DAT, &len);
	if (pb == NULL)
	{
		LOG_DEBUG("missing required field: STMP_TAG_DAT")
		stmpdec_free(&root);
		goto loop;
	}
	gt->begin = Gsc::newPbMsg(cb->begin, pb, len);
	if (gt->begin == NULL) /** 为简单起见(避免在业务处理中频繁判断BEGIN是否为空), BEGIN中的pb对象为必选. */
	{
		LOG_DEBUG("can not parse %s from STMP-BEGIN", it->second->begin->name().c_str())
		stmpdec_free(&root);
		goto loop;
	}
	uchar* att = stmpdec_peek_bin(&root, STMP_TAG_ATT, &len);
	if (att != NULL)
		gt->extbegin = new string((char*) att, len);
	stmpdec_free(&root);
	//
	if (Logger::isTrace())
	{
		LOG_TRACE("\n  <-- BEGIN(%s): {%s}, END(%s), UID: %llu, CMD: %04X, TID: %08X, PEER: %s\n", //
				cb->begin->name().c_str(), gt->begin == NULL ? "NULL" : gt->begin->ShortDebugString().c_str(),//
				cb->end->name().c_str(), this->uid, gt->cmd, gt->tid, Net::sockaddr2str(&gt->n2h->peer).c_str())
	}
	((void (*)(Gusr* gusr, Gn2htrans* gt, Message* begin)) (cb->cb))(this, gt, gt->begin);
}

/** 处理短连接上的STMP-BEGIN. */
void Gusr::procBegin1(uchar* dat, uint len, Gn2h* gt, ullong sid)
{
	uint out;
	uchar* msg = this->decryp(dat, len, &out); /** 将cid身后的字段解密. */
	stmp_node root;
	if (stmpdec_unpack(msg, out, &root) != 0)
	{
		loop: //
		gt->n2h->future([this, gt] /** 前往N2H关闭连接. */
		{
			gt->n2h->close();
			delete gt; /** 删除事务. */
		});
		free(msg);
		return;
	}
	Cb* cb = Gsc::unpkgBegin0(&root, &gt->tid, &gt->cmd, &gt->begin, &gt->extbegin);
	stmpdec_free(&root);
	if (cb == NULL || cb->cb == NULL)
	{
		LOG_DEBUG("it`s an unexpected cmd: %04X", gt->cmd)
		goto loop;
	}
	free(msg);
	//
	if (Logger::isTrace())
	{
		LOG_TRACE("\n  <-- BEGIN(%s): {%s}, END(%s), UID: %llu, SID: %llu, CMD: %04X, TID: %08X, PEER: %s\n", //
				cb->begin->name().c_str(), gt->begin == NULL ? "NULL" : gt->begin->ShortDebugString().c_str(),//
				cb->end->name().c_str(), this->uid, gt->sid, gt->cmd, gt->tid, Net::sockaddr2str(&gt->n2h->peer).c_str())
	}
	if (gt->sid != sid) /** 上行的会话ID与Gusr当前的csid不一致. */
	{
		LOG_DEBUG("session id does not match, may be have a new session created on this gusr, uid: %llu, gusr-sid: %llu, gt-sid: %llu", this->uid, sid, gt->sid)
		gt->end(RET_SID_ERROR);
		//lazy-close.
		return;
	}
	((void (*)(Gusr* gusr, Gn2h* gt, Message* begin)) (cb->cb))(this, gt, gt->begin);
}

/** 处理连接上的STMP-UNI, 只有长连接支持STMP-UNI. */
void Gusr::procUni(uchar* dat, uint len)
{
	uint out;
	uchar* msg = this->decryp(dat, len, &out); /** 将length身后的字段解密. */
	stmp_node root;
	if (stmpdec_unpack(msg, out, &root) != 0)
	{
		loop: //
		this->n2h->future([this] /** 前往N2H关闭连接. */
		{
			this->n2h->close();
		});
		free(msg);
		return;
	}
	ushort cmd;
	if (stmpdec_get_short(&root, STMP_TAG_CMD, &cmd) != 0)
	{
		LOG_DEBUG("missing required field: STMP_TAG_CMD")
		stmpdec_free(&root);
		goto loop;
	}
	auto it = Gsc::evns.find(cmd);
	if (it == Gsc::evns.end()) /** 找不到回调. */
	{
		LOG_WARN("can not found call back for this cmd: %04X", cmd); /* 不支持的命令字. */
		stmpdec_free(&root);
		goto loop;
	}
	Cb* cb = it->second;
	if (cb->uni == NULL || cb->cb == NULL)
	{
		LOG_DEBUG("it`s an unexpected cmd: %04X", cmd)
		stmpdec_free(&root);
		goto loop;
	}
	Message* uni = NULL;
	string* ext = NULL;
	uint size;
	uchar* pb = stmpdec_peek_bin(&root, STMP_TAG_DAT, &size);
	if (dat != NULL)
		uni = Gsc::newPbMsg(cb->uni, pb, len);
	uchar* att = stmpdec_peek_bin(&root, STMP_TAG_ATT, &size);
	if (att != NULL)
		ext = new string((char*) att, size);
	stmpdec_free(&root);
	free(msg);
	//
	LOG_TRACE("\n  <-- UNI(%s): {%s}, CMD: %04X, PEER: %s\n", cb->uni->name().c_str(), uni == NULL ? "NULL": uni->ShortDebugString().c_str(), cb->cmd, Net::sockaddr2str(&this->n2h->peer).c_str())
	((void (*)(Gusr* gusr, Message* uni, string* ext)) (cb->cb))(this, uni, ext);
	if (Gsc::ec)
		Gsc::ec->gn2hRecvUniOnGusr(this, cb->cmd, uni, ext);
	//
	if (uni != NULL)
		delete uni;
	if (ext != NULL)
		delete ext;
	return;
}

/** 加密, 并将事务前转到自身关联的N2H上出栈(长连接). */
void Gusr::sendEnd0(Gn2htrans* gt)
{
	if (Logger::isTrace())
	{
		auto it = Gsc::evns.find(gt->cmd);
		if (it != Gsc::evns.end())
		{
			Cb* cb = it->second;
			LOG_TRACE("\n  --> END(%s): {%s}, BEGIN(%s): {%s}, UID: %llu, RET: %04X, CMD: %04X, TID: %08X, PEER: %s\n", //
					cb->end->name().c_str(), gt->endx == NULL ? "NULL": gt->endx->ShortDebugString().c_str(),//
					cb->begin->name().c_str(), gt->begin == NULL ? "NULL": gt->begin->ShortDebugString().c_str(),//
					this->uid, gt->ret, gt->cmd, gt->tid, Net::sockaddr2str(&gt->n2h->peer).c_str())
		} else
		{
			LOG_FAULT("it`s a bug, cmd: %04X", gt->cmd)
		}
	}
	stmp_pdu* sp = Gsc::pkgEndSecDat(gt->ret, gt->endx, gt->extend);
	uint len;
	uchar* dat = stmpenc_take(sp, &len);
	uint out;
	uchar* cryp = this->encryp(dat, len, &out);
	free(sp->buff);
	free(sp);
	//
	stmp_pdu* end = (stmp_pdu*) malloc(sizeof(stmp_pdu));
	int size = out;
	size = size + (8 - (size % 8)) + 64; /** 只多不少. */
	end->len = size;
	end->rm = size;
	end->p = 0;
	end->buff = (uchar*) malloc(size);
	stmpenc_add_bin(end, STMP_TAG_DAT, cryp, out);
	stmpenc_add_int(end, STMP_TAG_DTID, gt->tid);
	stmpenc_add_tag(end, STMP_TAG_TRANS_END);
	dat = stmpenc_take(end, &len);
	//
	gt->n2h->future([gt, end, dat, len]
	{
		gt->n2h->send(dat, len);
		free(end->buff);
		free(end);
		gt->finish();
	});
	free(cryp);
}

/** 加密, 并将事务前转到自身关联的N2H上出栈(短连接). */
void Gusr::sendEnd1(Gn2h* gt)
{
	if (Logger::isTrace())
	{
		auto it = Gsc::evns.find(gt->cmd);
		if (it != Gsc::evns.end())
		{
			Cb* cb = it->second;
			LOG_TRACE("\n  --> END(%s): {%s}, BEGIN(%s): {%s}, UID: %llu, RET: %04X, CMD: %04X, TID: %08X, PEER: %s\n", //
					cb->end->name().c_str(), gt->endx == NULL ? "NULL": gt->endx->ShortDebugString().c_str(),//
					cb->begin->name().c_str(), gt->begin == NULL ? "NULL": gt->begin->ShortDebugString().c_str(),//
					this->uid, gt->ret, gt->cmd, gt->tid, Net::sockaddr2str(&gt->n2h->peer).c_str())
		} else
		{
			LOG_FAULT("it`s a bug, cmd: %04X", gt->cmd)
		}
	}
	stmp_pdu* sp = Gsc::pkgEndSecDat(gt->ret, gt->endx, gt->extend);
	uint len;
	uchar* dat = stmpenc_take(sp, &len);
	uint out;
	uchar* cryp = this->encryp(dat, len, &out); /* 加密. */
	free(sp->buff);
	free(sp);
	//
	stmp_pdu* end = (stmp_pdu*) malloc(sizeof(stmp_pdu));
	int size = out;
	size = size + (8 - (size % 8)) + 64; /** 只多不少. */
	end->len = size;
	end->rm = size;
	end->p = 0;
	end->buff = (uchar*) malloc(size);
	stmpenc_add_bin(end, STMP_TAG_DAT, cryp, out);
	stmpenc_add_int(end, STMP_TAG_DTID, gt->tid);
	stmpenc_add_tag(end, STMP_TAG_TRANS_END);
	dat = stmpenc_take(end, &len);
	//
	gt->n2h->future([gt, end, dat, len]
	{
		gt->n2h->send(dat, len);
		gt->finish();
		free(end->buff);
		free(end);
	});
	free(cryp);
}

/** 加密, 并将消息前转到自身关联的N2H上出栈. */
void Gusr::sendUni(ushort cmd, Message* uni, string* ext)
{
	if (Logger::isTrace())
	{
		auto it = Gsc::evns.find(cmd);
		if (it != Gsc::evns.end())
		{
			Cb* cb = it->second;
			LOG_TRACE("\n  --> UNI(%s): {%s}, UID: %llu, CMD: %04X, PEER: %s\n", //
					cb->uni->name().c_str(), uni == NULL ? "NULL" : uni->ShortDebugString().c_str(),//
					this->uid, cmd, this->n2h == NULL ? "NULL" : Net::sockaddr2str(&this->n2h->peer).c_str())
		} else
		{
			LOG_FAULT("it`s a bug, cmd: %04X", cmd)
		}
	}
	if (this->n2h == NULL)
	{
		if (Gsc::ec)
			Gsc::ec->gn2hSendUniOnGusr(this, cmd, uni, ext);
		return;
	}
	stmp_pdu* sp = this->pkgUni(cmd, uni, ext);
	this->n2h->future([this, sp]
	{
		uint len;
		uchar* dat = stmpenc_take(sp, &len);
		this->n2h->send(dat, len);
		free(sp->buff);
		free(sp);
	});
	if (Gsc::ec)
		Gsc::ec->gn2hSendUniOnGusr(this, cmd, uni, ext);
	if (uni != NULL)
		delete uni;
	if (ext != NULL)
		delete ext;
}

/** 网络连接是否正常. */
bool Gusr::isOnline()
{
	return this->n2h != NULL;
}

/** 踢下线(将导致连接被延迟关闭). */
void Gusr::kick(ushort cmd, Message* uni)
{
	stmp_pdu* sp = this->pkgUni(cmd, uni, NULL);
	N2H* n2h = this->n2h;
	n2h->future([n2h, sp]
	{
		uint len;
		uchar* dat = stmpenc_take(sp, &len);
		n2h->send(dat, len);
		free(sp->buff);
		free(sp);
		n2h->lts = DateMisc::getMsec() + Cfg::libgsc_n2h_lazy_close;
		n2h->lazyClose();
		if (n2h->gusr != NULL) /** 如果gusr不为NULL, 则一定存在相互引用. */
		{
			Gusr* gusr = n2h->gusr;
			gusr->future([gusr]
					{
						/** gusr->n2h = NULL; 由于这里是并行, 可能将老的连接连接踢下线前, gusr->n2h已经被新的N2H覆盖, 因此, 这里不能置为NULL. */
						gusr->unRef();
					});
			n2h->gusr = NULL; /** N2H亦不再与Gusr关联. */
			n2h->unRef();
		}
	});
	this->n2h = NULL; /** 连接置空. */
	if (Gsc::ec)
		Gsc::ec->gn2hSendUniOnGusr(this, cmd, uni, NULL);
	if (uni != NULL)
		delete uni;
}

/** 踢下线, 无消息提示(连接被立即关闭). */
void Gusr::kickNoMsg()
{
	N2H* n2h = this->n2h;
	n2h->future([n2h]
	{
		Lx::getGwk()->removeActorNet(n2h); /** 关闭连接. */
		/**
		 *
		 * 这里为了达到静默关闭目的, 没调用n2h->close();
		 *
		 * */
		if (n2h->gusr != NULL) /** 如果gusr不为NULL, 则一定存在相互引用. */
		{
			Gusr* gusr = n2h->gusr;
			gusr->future([gusr]
					{
						/** gusr->n2h = NULL; 由于这里是并行, 可能将老的连接连接踢下线前, gusr->n2h已经被新的N2H覆盖, 因此, 这里不能置为NULL. */
						gusr->unRef();
					});
			n2h->gusr = NULL; /** N2H亦不再与Gusr关联. */
			n2h->unRef();
		}
		n2h->del();
	});
	this->n2h = NULL; /** 连接置空. */
}

/** 构造一个加密过的UNI消息. */
stmp_pdu* Gusr::pkgUni(ushort cmd, Message* uni, string* ext)
{
	stmp_pdu* sp = Gsc::pkgUniSecDat(cmd, uni, ext);
	uint len;
	uchar* dat = stmpenc_take(sp, &len);
	uint out;
	uchar* cryp = this->encryp(dat, len, &out); /* 加密. */
	free(sp->buff);
	free(sp);
	//
	sp = (stmp_pdu*) malloc(sizeof(stmp_pdu));
	int size = out;
	size = size + (8 - (size % 8)) + 64; /** 只多不少. */
	sp->len = size;
	sp->rm = size;
	sp->p = 0;
	sp->buff = (uchar*) malloc(size);
	stmpenc_add_bin(sp, STMP_TAG_DAT, cryp, out);
	stmpenc_add_tag(sp, STMP_TAG_TRANS_UNI);
	free(cryp);
	return sp;
}

Gusr::~Gusr()
{

}

